tools/tungsten/core/src/com/google/android/filament/tungsten/model/Nodes.kt

```diff
@@ -43,27 +45,63 @@ private val adderNodeCompile = fun(node: Node, compiler: GraphCompiler): Node {
     return node
 }
 
+private val UNLIT_INPUTS = listOf("baseColor", "emissive")
+private val LIT_INPUTS = listOf("baseColor", "metallic", "roughness", "reflectance", "clearCoat",
+        "clearCoatRoughness", "ambientOcclusion", "emissive")
+
+private val inputSlotsForShadingModel = { shadingModel: String ->
+    when (shadingModel) {
+        "unlit" -> UNLIT_INPUTS
+        "lit" -> LIT_INPUTS
+        else -> emptyList()
+    }
+}
+
 private val shaderNodeCompile = fun(node: Node, compiler: GraphCompiler): Node {
-    val baseColor = compiler.compileAndRetrieveExpression(node.getInputSlot("baseColor"))
-    val metallic = compiler.compileAndRetrieveExpression(node.getInputSlot("metallic"))
-    val roughness = compiler.compileAndRetrieveExpression(node.getInputSlot("roughness"))
+    val shadingModel = (node.properties[0].value as StringValue).value
+    compiler.setShadingModel(shadingModel)
+
+    val compileMaterialInput = { name: String, dimensions: Int, shouldProvideDefault: Boolean ->
+        val inputSlot = node.getInputSlot(name)
+        val connectedExpression = compiler.compileAndRetrieveExpression(inputSlot)
+        val isConnected = connectedExpression != null
+        val expression = connectedExpression ?: Literal(dimensions)
+
+        // Conform the expression to match the dimensionality of the material input.
+        val conformedExpression = expression.conform(dimensions)
+
+        // Some inputs, like clearCoat, affect shader code generation if they're set in GLSL. For
+        // these, shouldProvideDefault is set to false so that no code is added to set the material
+        // input.
+        if (isConnected || shouldProvideDefault) {
+            compiler.addCodeToMaterialFunctionBody("material.$name = $conformedExpression;\n")
+        }
+        compiler.setExpressionForSlot(inputSlot, conformedExpression)
+    }
 
-    if (baseColor != null) {
-        compiler.addCodeToMaterialFunctionBody("material.baseColor.rgb = ${baseColor.rgb};\n")
-        compiler.setExpressionForSlot(node.getInputSlot("baseColor"), baseColor.rgb)
+    // Compile inputs common to all shading models.
+    val compileCommonInputs = {
+        compileMaterialInput("baseColor", 4, true)
+        compileMaterialInput("emissive", 4, true)
     }
 
-    if (metallic != null) {
-        compiler.addCodeToMaterialFunctionBody("material.metallic = ${metallic.r};\n")
-        compiler.setExpressionForSlot(node.getInputSlot("metallic"), metallic.r)
+    // Compile inputs unique to the "lit" shading model.
+    val compileLitInputs = {
+        compileMaterialInput("metallic", 1, true)
```

Comment:
I'd recommend you use an enum instead of a boolean, it makes the call sites easier to read